src/odm-collection.hpp

```diff
@@ -0,0 +1,727 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <iostream>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <mongocxx/collection.hpp>
+
+#include <bson_mapper/bson_streambuf.hpp>
+
+// A mock class for testing serialization w/o an actual BSON serializer class.
+// This class is templated, but requires the given type to have primitive fields a, b, and c.
+// This will be removed once cereal::BSONArchiver is merged in
+template <class T>
+class out_archiver_mock {
+   public:
+    out_archiver_mock(std::ostream& os) : _os(&os) {
+    }
+
+    void operator()(T obj) {
+        bsoncxx::builder::basic::document basic_builder{};
+        basic_builder.append(bsoncxx::builder::basic::kvp("a", obj.a));
+        basic_builder.append(bsoncxx::builder::basic::kvp("b", obj.b));
+        basic_builder.append(bsoncxx::builder::basic::kvp("c", obj.c));
+        auto v = basic_builder.view();
+        _os->write(reinterpret_cast<const char*>(v.data()), v.length());
+    }
+
+   private:
+    std::ostream* _os;
+};
+
+// A mock class for testing de-serialization w/o an actual BSON serializer class.
+// This class is templated, but requiers the given type to ahve primitize fields a, b, c.
+// This will be removed once cereal::BSONArchiver is merged in
+template <class T>
+class in_archiver_mock {
+   public:
+    in_archiver_mock(std::istream& is) : _is(&is) {
+    }
+
+    void operator()(T& obj) {
+        // get document size
+        int size = 0;
+        _is->read(reinterpret_cast<char*>(&size), 4);
+        uint8_t* data = new uint8_t[size];
+        // read BSON data into document view
+        _is->seekg(0);
+        _is->read(reinterpret_cast<char*>(data), size);
+        auto doc_view = bsoncxx::document::view(data, size);
+        // Fill object with fields
+        obj.a = doc_view["a"].get_int32();
+        obj.b = doc_view["b"].get_int32();
+        obj.c = doc_view["c"].get_int32();
+    }
+
+   private:
+    std::istream* _is;
+};
+
+/**
+ * Converts a serializable object into a BSON document value
+ * TODO This is not very clean and kind of inefficient. Maybe we should be passing references
+ * into
+ * the bson_ostream callback?
+ * @tparam T   A type that is serializable to BSON using a BSONArchiver.
+ * @param  obj A serializable object
+ * @return     A BSON document value representing the given object.
+ */
+template <class T>
+bsoncxx::document::value to_document(const T& obj) {
+    bsoncxx::document::value doc(nullptr, 0, [](uint8_t*) {});
```

Comment:
I spoke with Jason about this, and I think there are a couple alternatives. _This way is fine_ but here are some alternatives:

use optional (we could `using` it in to a stdx namespace from the bsconcxx polyfill). something like:
```c++
stdx::optional<bsoncxx::document::value> doc;
...
return doc.get();
```

using [std::aligned_storage](http://en.cppreference.com/w/cpp/types/aligned_storage) to create uninitialized storage for the doc. something like:
```c++
using bsoncxx::document::value;  // for brevity
typename std::aligned_storage<sizeof(value), alignof(value)>::type doc;
bson_mapper::bson_ostream bos([&doc](value v) { new(&doc) value(std::move(v)); });
...
return reinterpret_cast<value>(doc);
```